src/interpreter/s2s_forward_mode_ad.jl

```diff
@@ -0,0 +1,147 @@
+
+
+    # We don't have a hand-coded rule, so derived one.
+    lock(MOONCAKE_INFERENCE_LOCK)
+    try
+        # If we've already derived the OpaqueClosures and info, do not re-derive, just
+        # create a copy and pass in new shared data.
+        oc_cache_key = ClosureCacheKey(interp.world, (sig_or_mi, debug_mode))
+        if haskey(interp.oc_cache, oc_cache_key)
+            return _copy(interp.oc_cache[oc_cache_key])
+        else
+            # Derive forward-pass IR, and shove in a `MistyClosure`.
+            forward_ir = generate_forward_ir(interp, sig_or_mi; debug_mode)
+            fwd_oc = MistyClosure(forward_ir; do_compile=true)
+            raw_rule = DerivedFRule(fwd_oc)
+            rule = debug_mode ? DebugFRule(raw_rule) : raw_rule
+            interp.oc_cache[oc_cache_key] = rule
+            return rule
+        end
+    catch e
+        rethrow(e)
+    finally
+        unlock(MOONCAKE_INFERENCE_LOCK)
+    end
+end
+
+function generate_forward_ir(
+    interp::MooncakeInterpreter,
+    sig_or_mi;
+    debug_mode=false,
+    do_inline=true,
+)
+    # Reset id count. This ensures that the IDs generated are the same each time this
+    # function runs.
+    seed_id!()
+
+    # Grab code associated to the primal.
+    primal_ir, _ = lookup_ir(interp, sig_or_mi)
+
+    # Normalise the IR.
+    isva, spnames = is_vararg_and_sparam_names(sig_or_mi)
+    ir = normalise!(primal_ir, spnames)
+
+    fwd_ir = dualize_ir(ir)
+    opt_fwd_ir = optimise_ir!(fwd_ir; do_inline)
+    return opt_fwd_ir
+end
+
+function dualize_ir(ir::IRCode)
+    new_stmts_stmt = map(make_fwd_ad_stmt, ir.stmts.stmt)
+    new_stmts_type = map(dual_type, ir.stmts.type)
+    new_stmts_info = ir.stmts.info
+    new_stmts_line = ir.stmts.line
+    new_stmts_flag = ir.stmts.flag
+    new_stmts = CC.InstructionStream(
+        new_stmts_stmt,
+        new_stmts_type,
+        new_stmts_info,
+        new_stmts_line,
+        new_stmts_flag,
+    )
+    new_cfg = ir.cfg
+    new_linetable = ir.linetable
+    rule_type = Any
+    new_argtypes = convert(Vector{Any}, vcat(rule_type, map(make_fwd_argtype, ir.argtypes)))
+    new_meta = ir.meta
+    new_sptypes = ir.sptypes
+    return IRCode(new_stmts, new_cfg, new_linetable, new_argtypes, new_meta, new_sptypes)
+end
+
+make_fwd_argtype(::Type{P}) where {P} = dual_type(P)
+make_fwd_argtype(c::Core.Const) = Dual  # TODO: refine to type of const
+
+function make_fwd_ad_stmt(stmt::Expr)
+    interp = get_interpreter()  # TODO: pass it around
+    C = context_type(interp)
+    if isexpr(stmt, :invoke) || isexpr(stmt, :call)
+        mi = stmt.args[1]::Core.MethodInstance
+        sig = mi.specTypes
+        if is_primitive(C, sig)
+            shifted_args = map(stmt.args) do a
+                if a isa Core.Argument
+                    Core.Argument(a.n + 1)
+                else
+                    a
+                end
+            end
```

Comment:
```suggestion
            shifted_args = inc_args(stmt.args)
```
I think you can just do this -- see https://github.com/compintell/Mooncake.jl/blob/5397d284e89f08afbb22cce32483c1ada6baec10/src/interpreter/bbcode.jl#L892

It's aimed at nodes we see in `BBCode`, but it works fine for `Expr`s and `ReturnNode`s, so I think it should work fine here.